### PR TITLE
Improve aggregate options

### LINKED ISSUE #75: Treat event enumerable command return type as a special case
At the moment, if we have an aggregate set up like:

```
var model = new ModelBuilder()
    .Aggregate<MutableAggregate, object>(aggregate =>
    {
        aggregate
            .WithRaisedEventsFrom(commandReturnTypes =>
            {
                commandReturnTypes
                    // Here we have a command return which is an enumerable of the events which were raised
                    .CommandReturnType<IEnumerable<object>>()
                    .WithEventsFrom(x => x);
            })
            .ApplyEventsByConvention()
            .FromMethodName(nameof(MutableAggregate.Apply))
            .IncludeNonPublicMethods();
    })
    .Build();
```

And then execute the command as follows:

`var events = context.ExecuteCommand(x => x.MyCommandMethod("value")).ToList();`

Then the command method is invoked twice, due to materializing again with `ToList`. This can be worked around with:

`var events = context.ExecuteCommand(x => x.MyCommandMethod("value").ToList().AsEnumerable()).ToList();`

To properly support this scenario, we should ideally be treating event enumerable command returns as a special case, where the implementation of `IMutableCommandReturnType` prevents the returned enumerable from being enumerated twice. This can be achieved by returning the materialized list of events which are a result of applying those events.

Thinking something like the following might make sense:

```
var model = new ModelBuilder()
    .Aggregate<MutableAggregate, object>(aggregate =>
    {
        aggregate
            .WithRaisedEventsFrom(commandReturnTypes =>
            {
                // A builder method to handle this specific case, which under the hood creates
                // an implementation of IMutableCommandReturnType for handling IEnumerable<TEventBase>.
                commandReturnTypes.EnumerableOfEvents();
            })
            .ApplyEventsByConvention()
            .FromMethodName(nameof(MutableAggregate.Apply))
            .IncludeNonPublicMethods();
    })
    .Build();
```

### PR DESCRIPTION
**Pull request by daniel-smith**
- Simplify aggregate options by removing a level of nesting for command results
- Handle command returns of type `IEnumerable<TEventBase`> by default
- Fixes #75
- Add some documentation (not complete, but a starting point)

### CODE REVIEW

File: src/DomainBlocks.Core/ApplyEventsBehavior.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+namespace DomainBlocks.Core;
+
+public enum ApplyEventsBehavior
+{
+    DoNotApply,
```
Comment (carlosrfernandez) [inline src/DomainBlocks.Core/ApplyEventsBehavior.cs:5]: We should probably add a description to these.

File: src/DomainBlocks.Core/Builders/AggregateOptionsBuilderBase.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace DomainBlocks.Core.Builders;
+
+public interface IAggregateOptionsBuilder
+{
+    public IAggregateOptions Options { get; }
```
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Core/Builders/AggregateOptionsBuilderBase.cs]: Done

File: src/DomainBlocks.Core/ApplyEventsBehavior.cs (hunk 1 above)
Comment (daniel-smith) [reply src/DomainBlocks.Core/ApplyEventsBehavior.cs:9]: Done

File: src/DomainBlocks.Core/ImmutableCommandResultOptions.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    private ImmutableCommandResultOptions(
+        ImmutableCommandResultOptions<TAggregate, TEventBase, TCommandResult> copyFrom)
+    {
+        _eventsSelector = copyFrom._eventsSelector;
+        _updatedStateSelector = copyFrom._updatedStateSelector;
+    }
+
+    public Type ClrType => typeof(TCommandResult);
+
+    public TCommandResult Coerce(TCommandResult commandResult, IEnumerable<object> raisedEvents)
```
Comment (carlosrfernandez) [inline src/DomainBlocks.Core/ImmutableCommandResultOptions.cs:26]: Is this method used?

**Review by carlosrfernandez — DISMISSED** [review]
Just some minor comments here and there

File: src/DomainBlocks.Core/ImmutableCommandResultOptions.cs (hunk 3 above)
Comment (daniel-smith) [reply src/DomainBlocks.Core/ImmutableCommandResultOptions.cs]: No. Removed.

File: src/DomainBlocks.Core/Builders/MutableCommandResultOptionsBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace DomainBlocks.Core.Builders;
+
+public interface IMutableCommandResultOptionsApplyEventsBuilder
+{
+    /// <summary>
+    /// Specify to not apply the returned events to the aggregate. Use this option when the mutable aggregate's state
+    /// is updated by the command method. This option is the default behaviour.
+    /// </summary>
+    public void DoNotApplyEvents();
```
[the remark this answers is not in the document]
Comment (daniel-smith) [reply src/DomainBlocks.Core/Builders/MutableCommandResultOptionsBuilder.cs]: This method isn't even used. Removed.

**Review by carlosrfernandez — APPROVED** [review]

**Pull request merged**